assets/src/block-validation/use-validation-error-state-updates.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * External dependencies
+ */
+import { isEqual } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+import { useSelect, useDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { AMP_VALIDITY_REST_FIELD_NAME } from './constants';
+import { BLOCK_VALIDATION_STORE_KEY } from './store';
+
+/**
+ * Handles a case that was encountered where a validationError's `sources` was an object (with numeric keys).
+ *
+ * @param {Object} validationError
+ */
+export function convertErrorSourcesToArray( validationError ) {
+	if ( ! Array.isArray( validationError.error.sources ) ) {
```

Comment:
I don't know exactly why this happened, but I occasionally came across blocks whose error sources looked like this instead of the expected array: 

```JS
error: {
	sources: {
		0: 'error1',
		1: 'error2',
		3: 'error3',
	},
},
```

I didn't take the time to figure out why, but it happened pretty regularly.

Reply:
Oh yes, that would explain it.

I'm pretty sure the problem is here:

https://github.com/ampproject/amp-wp/blob/a0f75534c06e8a290a58818183004c94118e9671/includes/validation/class-amp-validation-manager.php#L1191

For example:

```
array_unique( [ 0 => 'foo', 1 => 'bar', 2 => 'bar', 3 => 'baz' ], SORT_REGULAR );
=> array(3) {
  [0]=> string(3) "foo"
  [1]=> string(3) "bar"
  [3]=> string(3) "baz"
}
```

Therefore, I suggest making this change instead of having `convertErrorSourcesToArray`:

```php
--- a/includes/validation/class-amp-validation-manager.php
+++ b/includes/validation/class-amp-validation-manager.php
@@ -1189,7 +1189,7 @@ static function ( $enqueued_script_source ) use ( $script_handle ) {
 			}
 		}
 
-		$sources = array_unique( $sources, SORT_REGULAR );
+		$sources = array_values( array_unique( $sources, SORT_REGULAR ) );
 
 		return $sources;
 	}
```

Reply:
The only problem with this is that, if I understand correctly, this `array_unique` is running when a URL is validated and its validation errors are saved. Adding this `array_values` would help with newly stored errors, but removing the JS function would still lead to problems when dealing with existing errors. To confirm, I tried it, and the issue came back with an existing error. This completely breaks the editor.

<img width="853" alt="Screen Shot 2020-12-23 at 9 47 30 AM" src="https://user-images.githubusercontent.com/9020968/103014377-4bdea400-4504-11eb-96db-5639654e6805.png">

Reply:
I made the `array_values` change in bbb0e495a but left `convertErrorSourcesToArray` in place for now.

Reply:
Good. That makes sense. Maybe then add a jsdoc comment to `convertErrorSourcesToArray` which explains that it is only needed due to cached results before bbb0e49 was applied. Eventually, after stale results have been removed, we can remove it in favor of just short-circuiting if `sources` is not an array.

Reply:
Added the comment in 761712561

Reply:
Aside: The fully JS error from Contact Form 7 which has the duplicates is:

```json
    {
        "data": {
            "code": "DISALLOWED_TAG",
            "node_attributes": {
                "id": "jquery-migrate-js",
                "src": "https://wordpressdev.lndo.site/core-dev/src/wp-includes/js/jquery/jquery-migrate.js?ver=__normalized__"
            },
            "node_name": "script",
            "node_type": 1,
            "parent_name": "head",
            "sources": {
                "0": {
                    "dependency_handle": "jquery-migrate",
                    "dependency_type": "script",
                    "file": "includes/controller.php",
                    "function": "wpcf7_do_enqueue_scripts",
                    "handle": "contact-form-7",
                    "hook": "wp_enqueue_scripts",
                    "line": 21,
                    "name": "contact-form-7",
                    "priority": 10,
                    "type": "plugin"
                },
                "1": {
                    "dependency_handle": "jquery-migrate",
                    "dependency_type": "script",
                    "file": "script-loader.php",
                    "function": "wp_enqueue_scripts",
                    "handle": "contact-form-7",
                    "hook": "wp_head",
                    "line": 1998,
                    "name": "wp-includes",
                    "priority": 1,
                    "type": "core"
                },
                "10": {
                    "dependency_type": "script",
                    "file": "class-wp-widget-factory.php",
                    "function": "WP_Widget_Factory::_register_widgets",
                    "handle": "jquery-migrate",
                    "hook": "widgets_init",
                    "line": 89,
                    "name": "wp-includes",
                    "priority": 100,
                    "type": "core"
                },
                "12": {
                    "dependency_type": "script",
                    "file": "widgets.php",
                    "function": "wp_widgets_init",
                    "handle": "jquery-migrate",
                    "hook": "init",
                    "line": 1747,
                    "name": "wp-includes",
                    "priority": 1,
                    "type": "core"
                },
                "14": {
                    "file": "script-loader.php",
                    "function": "wp_print_head_scripts",
                    "hook": "wp_head",
                    "line": 1952,
                    "name": "wp-includes",
                    "priority": 9,
                    "type": "core"
                },
                "2": {
                    "dependency_handle": "jquery-migrate",
                    "dependency_type": "script",
                    "file": "script-loader.php",
                    "function": "wp_default_scripts",
                    "handle": "jquery",
                    "hook": "wp_default_scripts",
                    "line": 594,
                    "name": "wp-includes",
                    "priority": 10,
                    "type": "core"
                },
                "4": {
                    "dependency_handle": "jquery-migrate",
                    "dependency_type": "script",
                    "file": "class-wp-widget-factory.php",
                    "function": "WP_Widget_Factory::_register_widgets",
                    "handle": "jquery",
                    "hook": "widgets_init",
                    "line": 89,
                    "name": "wp-includes",
                    "priority": 100,
                    "type": "core"
                },
                "6": {
                    "dependency_handle": "jquery-migrate",
                    "dependency_type": "script",
                    "file": "widgets.php",
                    "function": "wp_widgets_init",
                    "handle": "jquery",
                    "hook": "init",
                    "line": 1747,
                    "name": "wp-includes",
                    "priority": 1,
                    "type": "core"
                },
                "8": {
                    "dependency_type": "script",
                    "file": "script-loader.php",
                    "function": "wp_default_scripts",
                    "handle": "jquery-migrate",
                    "hook": "wp_default_scripts",
                    "line": 594,
                    "name": "wp-includes",
                    "priority": 10,
                    "type": "core"
                }
            },
            "type": "js_error"
        },
        "term_slug": "ee3a2c7eb6a0a67be7e6b3cbf062901c"
    },
```

Reply:
For the record, this has been followed up in #5741 via a91b14a870b7669f1d0cdccb0dd10708ebc463f5.